src/java.base/share/classes/java/io/ProxyingConsole.java

```diff
@@ -0,0 +1,197 @@
+
+    ProxyingConsole(JdkConsole delegate) {
+        this.delegate = delegate;
+        readLock = new Object();
+        writeLock = new Object();
+        reader = new WrappingReader(delegate.reader(), readLock);
+        printWriter = new WrappingWriter(delegate.writer(), writeLock);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public PrintWriter writer() {
+        return printWriter;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Reader reader() {
+        return reader;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Console format(String fmt, Object ... args) {
+        synchronized (writeLock) {
+            delegate.format(fmt, args);
+        }
+        return this;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Console printf(String format, Object ... args) {
+        synchronized (writeLock) {
+            delegate.printf(format, args);
+        }
+        return this;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String readLine(String fmt, Object ... args) {
+        synchronized (writeLock) {
+            synchronized (readLock) {
+                return delegate.readLine(fmt, args);
+            }
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String readLine() {
+        synchronized (readLock) {
+            return delegate.readLine();
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public char[] readPassword(String fmt, Object ... args) {
+        synchronized (writeLock) {
+            synchronized (readLock) {
+                return delegate.readPassword(fmt, args);
+            }
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public char[] readPassword() {
+        synchronized (readLock) {
+            return delegate.readPassword();
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void flush() {
+        delegate.flush();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Charset charset() {
+        return delegate.charset();
+    }
+
+    private static class WrappingReader extends Reader {
+        private final Reader r;
+        private final Object lock;
+
+        WrappingReader(Reader r, Object lock) {
+            this.r = r;
+            this.lock = lock;
+        }
+
+        @Override
+        public int read(char[] cbuf, int off, int len) throws IOException {
+            synchronized (lock) {
+                return r.read(cbuf, off, len);
+            }
+        }
+
+        @Override
+        public void close() throws IOException {
+            r.close();
+        }
+    }
+
+    private static class WrappingWriter extends PrintWriter {
+        private final PrintWriter pw;
+        private final Object lock;
+
+        public WrappingWriter(PrintWriter pw, Object lock) {
+            super(pw);
```

Comment:
PrintWriter doesn't provide a way to provide to lock object so this means the overridden methods will synchronize on "lock", the non-overridden methods will synchronize on "pw". So we will need to look at this.